src/Cake.Core/CakeTask.cs

```diff
@@ -35,165 +30,114 @@ public abstract class CakeTask : ICakeTaskInfo
         /// Gets the task's dependencies.
         /// </summary>
         /// <value>The task's dependencies.</value>
-        public IReadOnlyList<CakeTaskDependency> Dependencies => _dependencies;
+        public List<CakeTaskDependency> Dependencies { get; }
 
         /// <summary>
         /// Gets the tasks that the task want to be a dependency of.
         /// </summary>
         /// <value>The tasks that the task want to be a dependency of.</value>
-        public IReadOnlyList<CakeTaskDependency> Dependees => _reverseDependencies;
+        public List<CakeTaskDependency> Dependees { get; }
 
         /// <summary>
         /// Gets the task's criterias.
         /// </summary>
         /// <value>The task's criterias.</value>
-        public IReadOnlyList<Func<ICakeContext, bool>> Criterias => _criterias;
+        public List<CakeTaskCriteria> Criterias { get; }
 
         /// <summary>
-        /// Gets the error handler.
+        /// Gets or sets the error handler.
         /// </summary>
         /// <value>The error handler.</value>
-        public Action<Exception> ErrorHandler { get; private set; }
+        public Action<Exception> ErrorHandler { get; set; }
 
         /// <summary>
-        /// Gets the error reporter.
+        /// Gets or sets the error reporter.
         /// </summary>
-        public Action<Exception> ErrorReporter { get; private set; }
+        public Action<Exception> ErrorReporter { get; set; }
 
         /// <summary>
-        /// Gets the finally handler.
+        /// Gets or sets the finally handler.
         /// </summary>
-        public Action FinallyHandler { get; private set; }
+        public Action FinallyHandler { get; set; }
 
         /// <summary>
-        /// Initializes a new instance of the <see cref="CakeTask"/> class.
+        /// Gets the task's actions.
         /// </summary>
-        /// <param name="name">The name of the task.</param>
-        protected CakeTask(string name)
-        {
-            if (name == null)
-            {
-                throw new ArgumentNullException(nameof(name));
-            }
-            if (string.IsNullOrWhiteSpace(name))
-            {
-                throw new ArgumentException("Task name cannot be empty.");
-            }
-
-            _dependencies = new List<CakeTaskDependency>();
-            _reverseDependencies = new List<CakeTaskDependency>();
-            _criterias = new List<Func<ICakeContext, bool>>();
-
-            Name = name;
-        }
+        /// <value>The task's actions.</value>
+        public List<Func<ICakeContext, Task>> Actions { get; }
```

Comment:
`IList`... Or is `ISet` or `ICollection` sufficient?

Reply:
We could change this to IList, but the collection should keep the order items were added and expose them by index to ensure backwards compatibility.